Fix wsgi environment building

redixin opened the pull request:
Fixed issues:

SERVER_NAME and SERVER_PORT variables may be determined in two ways:

 if protocol is HTTP/1.0:
  Value of `Host` header. If there is no such header, then address of
  transport may be taken (transport.get_extra_info('sockname'))

 if protocol is HTTP/1.1:
  Value of `Host` header. If there is no such header, the Bad Request
  should be raised, because this header is required in HTTP/1.1.

REMOTE_ADDR and REMOTE_PORT should be address and port of host connected
to http server.

```
https://www.ietf.org/rfc/rfc3875:
4.1.8.  REMOTE_ADDR
   The REMOTE_ADDR variable MUST be set to the network address of the
   client sending the request to the server.
```

Also, header `Authorization` should not be in environment at all.

redixin commented on tests/test_wsgi.py:
```diff
@@ -280,3 +253,20 @@ def test_dont_unquote_environ_path_info(self):
+            'GET', '/', (1, 0), headers, True, 'deflate')
+        environ = self._make_one()
+        self.assertEqual('1.2.3.4', environ["SERVER_NAME"])
+        self.assertEqual('8080', environ["SERVER_PORT"])
```
Hmmm probably I should also check self.transport.get_extra_info called with proper arguments...

asvetlov commented:
@redixin please add a message next time after pushing an update for PR.
The reason is: github did not send a notification email on pushes, only on new comments.
I rely on these emails to track PR activities.

asvetlov commented on aiohttp/wsgi.py:
```diff
@@ -63,16 +63,11 @@ def create_wsgi_environ(self, message, payload):
             'SERVER_PROTOCOL': 'HTTP/%s.%s' % message.version
         }
 
-        # authors should be aware that REMOTE_HOST and REMOTE_ADDR
```
Maybe the comment worth to be saved?
What do you think?

redixin replied:
Sure. Added this comment on lines 86-90

asvetlov commented on tests/test_wsgi.py:
```diff
@@ -75,10 +74,9 @@ def test_environ_headers(self):
         self.assertEqual(environ['SCRIPT_NAME'], 'script')
         self.assertEqual(environ['SERVER_NAME'], 'python.org')
-        self.assertEqual(environ['SERVER_PORT'], '443')
+        self.assertEqual(environ['SERVER_PORT'], '80')
```
I don't follow why the port has changed?

redixin replied:
Port has changed because old version was trying to determine port number by url scheme. (see line 71 is_ssl=True)

So in old version it was '80' if url_scheme == 'http' else '443' (see old version aiohttp/wsgi.py:95-98)

But there is possible situations:

 http://example.com:8080/
 https://example.com:4443/

Old version will give wrong result in both cases.

Ill add more tests to cover it.

redixin replied:
Oops. I just realized that client may connect to port 123 but say
```GET / HTTP/1.1
Host: example.com:456
```
According to CGI 1.1 in this case SERVER_PORT should be 123.

Ill fix it right now.

redixin commented:
New patch added
```
    Fix SERVER_PORT in wsgi headers
    
    According to CGI 1.1 server port should be set to the TCP/IP port
    number on which this request is received from the client.
    
    Also add comment about REMOTE_HOST, REMOTE_ADDR and SERVER_PORT.
    
    Also added new and removed some redundant tests.
```

asvetlov commented:
Perfect!
Thanks a lot!
